src/driver/sqljs/SqljsQueryRunner.ts

```diff
@@ -0,0 +1,117 @@
+import {ObjectLiteral} from "../../common/ObjectLiteral";
+import {QueryRunnerAlreadyReleasedError} from "../../error/QueryRunnerAlreadyReleasedError";
+import {OrmUtils} from "../../util/OrmUtils";
+import {InsertResult} from "../InsertResult";
+import {AbstractSqliteQueryRunner} from "../sqlite-abstract/AbstractSqliteQueryRunner";
+import {SqljsDriver} from "./SqljsDriver";
+
+/**
+ * Runs queries on a single sqlite database connection.
+ *
+ * Does not support compose primary keys with autoincrement field.
+ * todo: need to throw exception for this case.
+ */
+export class SqljsQueryRunner extends AbstractSqliteQueryRunner {
+    
+    /**
+     * Database driver used by connection.
+     */
+    driver: SqljsDriver;
+    
+    // -------------------------------------------------------------------------
+    // Constructor
+    // -------------------------------------------------------------------------
+
+    constructor(driver: SqljsDriver) {
+        super(driver);
+        this.driver = driver;
+        this.connection = driver.connection;
+    }
+
+    // -------------------------------------------------------------------------
+    // Public methods
+    // -------------------------------------------------------------------------
+
+    /**
+     * Executes a given SQL query.
+     */
+    query(query: string, parameters?: any[]): Promise<any> {
+        if (this.isReleased)
+            throw new QueryRunnerAlreadyReleasedError();
+
+        return new Promise<any[]>(async (ok, fail) => {
+            const databaseConnection = await this.connect();
+            this.driver.connection.logger.logQuery(query, parameters, this);
+            const queryStartTime = +new Date();
+            try {
+                const statement = databaseConnection.prepare(query);
+                statement.bind(parameters);
+                
+                // log slow queries if maxQueryExecution time is set
+                const maxQueryExecutionTime = this.driver.connection.options.maxQueryExecutionTime;
+                const queryEndTime = +new Date();
+                const queryExecutionTime = queryEndTime - queryStartTime;
+                if (maxQueryExecutionTime && queryExecutionTime > maxQueryExecutionTime)
+                    this.driver.connection.logger.logQuerySlow(queryExecutionTime, query, parameters, this);
+
+                const result: any[] = [];
+
+                while (statement.step()) {
+                    result.push(statement.getAsObject());
+                }
+                
+                if (query.toUpperCase().match(/(COMMIT)/)) {
+                    this.driver.autoSave();
```

Comment:
what if `autoSave` is an async operation? Callback provided by a user most probably should be a promise since it'll most probably save in the cloud / filesystem / database using async api

Reply:
also this code looks a bit like a monkey patch. Isn't it better to do it after `save` method call? Plus after operations called by insert/update/delete query builders?

Reply:
Yes, it's a bit of a monkey patch because I didn't want to save after every update/insert/delete, because they are always executed in a transaction, right? So I guess the best place would be `commitTransaction()`.

> what if autoSave is an async operation?

The question is, do we have to wait for the save operation to complete? What would change?